bitsail-connectors/connector-ftp/src/main/java/com/bytedance/bitsail/connector/ftp/source/split/coordinator/FtpSourceSplitCoordinator.java

```diff
@@ -0,0 +1,176 @@
+    this.context = context;
+    this.jobConf = jobConf;
+    this.splitAssignmentPlan = Maps.newConcurrentMap();
+  }
+
+  @Override
+  public void start() {
+    List<FtpSourceSplit> splitList;
+    try {
+      IFtpHandler ftpHandler = FtpHandlerFactory.createFtpHandler(new FtpConfig(jobConf));
+      splitList = construct(ftpHandler);
+    } catch (IOException e) {
+      throw new BitSailException(FtpErrorCode.SPLIT_ERROR, "Failed to create splits.");
+    }
+
+    int readerNum = context.totalParallelism();
+    LOG.info("Found {} readers and {} splits.", readerNum, splitList.size());
+    if (readerNum > splitList.size()) {
+      LOG.error("Reader number {} is larger than split number {}.", readerNum, splitList.size());
+    }
+
+    for (FtpSourceSplit split : splitList) {
+      int readerIndex = ReaderSelector.getReaderIndex(readerNum);
+      splitAssignmentPlan.computeIfAbsent(readerIndex, k -> new HashSet<>()).add(split);
+      LOG.info("Will assign split {} to the {}-th reader", split.uniqSplitId(), readerIndex);
+    }
+  }
+
+  @Override
+  public void addReader(int subtaskId) {
+    LOG.info("Found reader {}", subtaskId);
+    tryAssignSplitsToReader();
+  }
+
+  @Override
+  public void addSplitsBack(List<FtpSourceSplit> splits, int subtaskId) {
+    LOG.info("Source reader {} return splits {}.", subtaskId, splits);
+
+    int readerNum = context.totalParallelism();
+    for (FtpSourceSplit split : splits) {
+      int readerIndex = ReaderSelector.getReaderIndex(readerNum);
+      splitAssignmentPlan.computeIfAbsent(readerIndex, k -> new HashSet<>()).add(split);
+      LOG.info("Re-assign split {} to the {}-th reader.", split.uniqSplitId(), readerIndex);
+    }
+
+    tryAssignSplitsToReader();
+  }
+
+  private void tryAssignSplitsToReader() {
+    Map<Integer, List<FtpSourceSplit>> splitsToAssign = new HashMap<>();
+
+    for (Integer readerIndex : splitAssignmentPlan.keySet()) {
+      if (CollectionUtils.isNotEmpty(splitAssignmentPlan.get(readerIndex)) && context.registeredReaders().contains(readerIndex)) {
+        splitsToAssign.put(readerIndex, Lists.newArrayList(splitAssignmentPlan.get(readerIndex)));
+      }
+    }
+
+    for (Integer readerIndex : splitsToAssign.keySet()) {
+      LOG.info("Try assigning splits reader {}, splits are: [{}]", readerIndex,
+              splitsToAssign.get(readerIndex).stream().map(FtpSourceSplit::getSplitId).collect(Collectors.toList()));
+      splitAssignmentPlan.remove(readerIndex);
+      context.assignSplit(readerIndex, splitsToAssign.get(readerIndex));
+      context.signalNoMoreSplits(readerIndex);
+      LOG.info("Finish assigning splits reader {}", readerIndex);
+    }
+  }
+
+  private List<FtpSourceSplit> construct(IFtpHandler ftpHandler) throws IOException {
+    List<FtpSourceSplit> splits = new ArrayList<>();
+    ftpHandler.loginFtpServer();
+    String[] paths = ftpHandler.getFtpConfig().getPaths();
```

Comment:
will we support regex in the future?

Reply:
I added todo list for this, and I will try to implement it in another pr.